### PR TITLE
[MIG] web_environment_ribbon: Migration to 15.0

### PR DESCRIPTION
**Pull request by ap-wtioit**
replaces #2049, now based on branch 15.0 and following https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0

Info @wt-io-it

### CODE REVIEW

**Comment by ap-wtioit** [comment]
Couldn't trigger a proper run in runbot. So i guess now it's time to wait until runbot works with 15.0.

**Comment by leemannd** [comment]
`First-time contributors need a maintainer to approve running workflows. Learn more. 1 pending check`
It wasn't present before. We will need help from @OCA/website-maintainers

**Comment by pedrobaeza** [comment]
AFAIK, runbot is not yet working for v15, due to change to GitHub actions and the dependency to travis2docker thing.

**Comment by leemannd** [comment]
Hello @pedrobaeza , it is launched on other repo:
- https://github.com/OCA/server-tools/pull/2193 This one worked out of the box
- https://github.com/OCA/server-env/pull/83 This one required a manual operation

**Comment by pedrobaeza** [comment]
I'm afraid not, Denis. Enter into "Details", and you'll see "Skipped" status due to what I have commented.

**Comment by leemannd** [comment]
I should have been reading better. Thank you Perdo for the informations and the time taken.

**Comment by ap-wtioit** [comment]
For the pre-commit issue `distutils.errors.DistutilsSetupError: Unsupported odoo version '15.0' in ./web_environment_ribbon` i opened https://github.com/OCA/maintainer-tools/issues/513

**Comment by ap-wtioit** [comment]
rebased to include a652ba3dfdf6d887e9f845f56741e00aa2d4ac2d with newer setuptools-odoo that support Odoo 15.0

File: web_environment_ribbon/tests/test_environment_ribbon_data.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2019 Eric Lembregts
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.tests import common
+
+
+class TestEnvironmentRibbonData(common.SavepointCase):
```
Comment (simahawk) [inline web_environment_ribbon/tests/test_environment_ribbon_data.py]: SavepointCase is deprecated

**Review by simahawk — APPROVED** [review]
LG, code review only

File: web_environment_ribbon/tests/test_environment_ribbon_data.py (hunk 1 above)
Comment (ap-wtioit) [reply web_environment_ribbon/tests/test_environment_ribbon_data.py]: thanks for the hint, i did miss this update in the migration guide

**Review by leemannd — APPROVED** [review]
Seems good. Thanks

**Comment by ph-obs** [comment]
Do you guys know when this migration will be finished? Thanks

**Review by wtaferner — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Very weird. Let's try again: /ocabot merge nobump

**Comment by pedrobaeza** [comment]
I will try tomorrow to see if it's something sporadic

**Comment by hbrunn** [comment]
or ocabot just hates Pedro? So with another account and later /ocabot merge nobump

**Comment by hbrunn** [comment]
not sporadic it seems or i'm also in disgrace of ocabot

**Comment by pedrobaeza** [comment]
I have just merged other PR in this repo, so this seems something specific to this PR. @ap-wtioit maybe you can rebase to see what's happening?

**Comment by ap-wtioit** [comment]
It looks to me as if the CI has some bugs (https://github.com/OCA/web/runs/4193888555):
```
+ pip install -r test-requirements.txt
-e ./setup/web_environment_ribbon
WARNING: The directory '/github/home/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
ERROR: ./setup/web_environment_ribbon is not a valid editable requirement. It should either be a path to a local project or a VCS URL (beginning with bzr+http, bzr+https, bzr+ssh, bzr+sftp, bzr+ftp, bzr+lp, bzr+file, git+http, git+https, git+ssh, git+git, git+file, hg+file, hg+http, hg+https, hg+ssh, hg+static-http, svn+ssh, svn+http, svn+https, svn+svn, svn+file).
Error: Process completed with exit code 1.
```

Will rebase, but i already did that a few weeks ago and the CI not starting for new users, then waiting a few weeks (hopefully not months) for somebody to test it is very anoying.

**Comment by ap-wtioit** [comment]
Better error messages would be needed i think:
```
+ pip install -r test-requirements.txt
ERROR: ./setup/web_environment_ribbon is not a valid editable requirement. It should either be a path to a local project or a VCS URL (beginning with bzr+http, bzr+https, bzr+ssh, bzr+sftp, bzr+ftp, bzr+lp, bzr+file, git+http, git+https, git+ssh, git+git, git+file, hg+file, hg+http, hg+https, hg+ssh, hg+static-http, svn+ssh, svn+http, svn+https, svn+svn, svn+file).
```
Somehow pre-commit did not create the setup directory right after the 15.0 release (when this was migrated)

**Comment by pedrobaeza** [comment]
But the folder is created by pre-commit itself. Did you run it?

**Comment by ap-wtioit** [comment]
@pedrobaeza i ran it when migrating (there where bugs in the OCA tooling then) i ran it again, but now i get this error:
```
+ export SETUPTOOLS_ODOO_POST_VERSION_STRATEGY_OVERRIDE=none
+ SETUPTOOLS_ODOO_POST_VERSION_STRATEGY_OVERRIDE=none
+ touch test-requirements.txt
++ addons --addons-dir /mnt/data/odoo-data-dir --include '' --exclude '' --separator ' ' list
+ for addon in $(addons --addons-dir "${ADDONS_DIR}" --include "${INCLUDE}" --exclude "${EXCLUDE}" --separator " " list)
+ echo '-e ./setup/web_environment_ribbon'
+ for addon in $(addons --addons-dir "${ADDONS_DIR}" --include "${INCLUDE}" --exclude "${EXCLUDE}" --separator " " list)
+ echo '-e ./setup/web_m2x_options'
+ cat test-requirements.txt
-e ./setup/web_environment_ribbon
-e ./setup/web_m2x_options
+ pip install -r test-requirements.txt
Looking in indexes: https://wheelhouse.odoo-community.org/oca-simple-and-pypi
Obtaining file:///mnt/data/odoo-data-dir/setup/web_environment_ribbon (from -r test-requirements.txt (line 1))
    ERROR: Command errored out with exit status 1:
     command: /opt/odoo-venv/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/mnt/data/odoo-data-dir/setup/web_environment_ribbon/setup.py'"'"'; __file__='"'"'/mnt/data/odoo-data-dir/setup/web_environment_ribbon/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-umhce5zy
         cwd: /mnt/data/odoo-data-dir/setup/web_environment_ribbon/
    Complete output (23 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/mnt/data/odoo-data-dir/setup/web_environment_ribbon/setup.py", line 3, in <module>
        setuptools.setup(
      File "/opt/odoo-venv/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
        return distutils.core.setup(**attrs)
      File "/usr/lib/python3.8/distutils/core.py", line 108, in setup
        _setup_distribution = dist = klass(attrs)
      File "/opt/odoo-venv/lib/python3.8/site-packages/setuptools/dist.py", line 453, in __init__
        _Distribution.__init__(
      File "/usr/lib/python3.8/distutils/dist.py", line 292, in __init__
        self.finalize_options()
      File "/opt/odoo-venv/lib/python3.8/site-packages/setuptools/dist.py", line 831, in finalize_options
        ep(self)
      File "/opt/odoo-venv/lib/python3.8/site-packages/setuptools/dist.py", line 852, in _finalize_setup_keywords
        ep.load()(self, ep.name, value)
      File "/opt/odoo-venv/lib/python3.8/site-packages/setuptools_odoo/setup_keywords.py", line 66, in odoo_addon
        setup_keywords = prepare_odoo_addon(
      File "/opt/odoo-venv/lib/python3.8/site-packages/setuptools_odoo/core.py", line 529, in prepare_odoo_addon
        addons_dir, addons_ns = _find_addons_dir()
      File "/opt/odoo-venv/lib/python3.8/site-packages/setuptools_odoo/core.py", line 335, in _find_addons_dir
        raise RuntimeError("No addons namespace found.")
    RuntimeError: No addons namespace found.
```

Edit: 
setup/web_environment_ribbon/odoo/addons/web_environment_ribbon symlink was not commited

**Comment by pedrobaeza** [comment]
I didn't get that error, so your pre-commit local installation should be damage. Sometimes `pre-commit clean` fixes this. Anyways, I see that you have already found the current problem, and I think it should be OK now.

/ocabot merge nobump

**Comment by ap-wtioit** [comment]
runbot still says no and gives no error message. 

https://runbot.odoo-community.org/runbot/static/build/3516324-2055-387d45/logs/job_10_test_base.txt:
```
Step 21/22 : RUN /bin/bash -c "source ${REPO_REQUIREMENTS}/virtualenv/python3.8/bin/activate && source ${REPO_REQUIREMENTS}/virtualenv/nodejs/bin/activate && source /rvm_env.sh && /install"
 ---> Running in 816998b914d0
[91m/bin/bash: /.repo_requirements/virtualenv/python3.8/bin/activate: No such file or directory
```

https://runbot.odoo-community.org/runbot/static/build/3516324-2055-387d45/logs/job_20_test_all.txt
```
Unable to find image 'oca-web:387d45e_env_1_job_1' locally
docker: Error response from daemon: pull access denied for oca-web, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
```

Edit:
Runbot started this image: https://hub.docker.com/r/vauxoo/odoo-80-image-shippable-auto for 15.0 migration branch
Documentation says this is only suitable up to version 10.0
```
Base docker image for Odoo 6.1, 7.0, 8.0,9.0 and 10.0 instances
```
Should have been this image:
https://hub.docker.com/r/vauxoo/odoo-150-image

Where should i report this?

**Comment by pedrobaeza** [comment]
Runbot is not working for v15 and its replacement is runboat (note the `a` in between and the word game), but all of this is still a bit experimental and not officially communicated. In a nutshell, it's because the runbot we have runs at maximum Python 3.7, and Odoo 15 is working now with 3.8.

**Comment by ap-wtioit** [comment]
@pedrobaeza thanks for your support in this one.

**Comment by pedrobaeza** [comment]
@ap-wtioit it seems there's a glitch in the module: when you hover on the navbar, the ribbon disappears. Have you experimented it?

**Comment by ap-wtioit** [comment]
@pedrobaeza that's this commit:
https://github.com/OCA/web/pull/2055/commits/fcb8c6e5a5e799225b49c73a8f4e999f9fab30e4
Otherwise the banner blocks the view for the first menu entry and the app menu.

**Comment by pedrobaeza** [comment]
Uhm, I see it as a weird design decision. Previously there wasn't this problem, making the ribbon transparent and clickable-through it. Can this behavior be put now?

**Comment by ap-wtioit** [comment]
It's still click through and transparent. But it visually blocked my intention of clicking the app menu. When migrating i felt as if the banner blocked my view and i didn't expect that something that is hidden to be clickable.

image

So for me the fix was that the banner get's out of the way if i move the cursor to the menu.

**Comment by pedrobaeza** [comment]
But following your own picture, that's not happening:

151122485-ced6c36c-76df-4e8b-99cb-26c75b67419c

I see this behavior as very confusing and visually not ergonomic, as well as not consistent with previous versions. Can you please revert back such change?

**Comment by ap-wtioit** [comment]
i took the screenshot without the cursor hovering the menu. With hovering the menu the banner becomes invisible and the user can navigate without the banner in the way:
image

Why did you mark the main Apps menu entry as not clickable in your screenshot?

**Comment by pedrobaeza** [comment]
> Why did you mark the main Apps menu entry as not clickable in your screenshot?

Sorry, my mistake, but even being clickable, it's not the usual thing.

The question remains the same: hiding a big element like the ribbon when moving the cursor is not visually ergonomic, provoking eyestrain and even can be flashing epileptic warning. And it's happening on the whole navbar, even on the farthest right part. I beg you to reconsider this behavior and revert back it.

**Comment by ap-wtioit** [comment]
Hiding an element when moving the cursor seems like an ok pattern to me. Regarding the flashing i can see that issue and could check for a solution.

Reverting the change can be done by anyone but it feels a bit strange to me that you seem to force me to do so.

To be clear the old implementation made it quite impossible to read some of the stuff hidden behind the banner so i tried to fix that. I still would want to work together for a solution for this issue that you cannot read stuff behind the banner and a new user would not expect it to be clickable (because visually it's not):
image

**Comment by pedrobaeza** [comment]
I don't want to force you, and of course I can propose the revert myself. Sorry if you have seen my comments as a command. What I don't want is to propose such revert and don't agree on doing such move.

About hiding the apps dropdown, I don't see it as a big deal, as 99% of people will use either `web_responsive` or `web_enterprise` modules, so that dropdown is replaced by the icons view, and thus, not overlapping the ribbon.

Anyways, other option can be to change z-index for putting such dropdown on top of the ribbon.

**Comment by ap-wtioit** [comment]
I got that it's not a big deal for you to hide all the items behind the banner, as i would expect you to know Odoo from the inside out.

I'm trying to make Odoo more user friendly for the people new to Odoo. And reading the menus and the headings is one of the first things a user should do to orient themselves within a new software. We also teach users to use the test system first (where we use the banner). The Banner makes all the Text behind it unreadable in terms of contrast ratios. So i tried to fix that with the current commit (that can be improved). [Based on the correlation of eye movement to the cursor.](https://doi.org/10.1145/634067.634234) I would like to work with you on improvements for the issues you have. But i think it would be best done in a new issue or pull request where we can discuss improvements.

**Pull request merged**